keyboards/snes_macropad/snes_macropad.c

```diff
@@ -0,0 +1,129 @@
+
+static void set_keylog(uint16_t keycode, keyrecord_t *record) {
+    key_name     = ' ';
+    last_keycode = keycode;
+    if (IS_QK_MOD_TAP(keycode)) {
+        if (record->tap.count) {
+          keycode = QK_MOD_TAP_GET_TAP_KEYCODE(keycode);
+        } else {
+          keycode = 0xE0 + biton(QK_MOD_TAP_GET_MODS(keycode) & 0xF) + biton(QK_MOD_TAP_GET_MODS(keycode) & 0x10);
+        }
+    } else if (IS_QK_LAYER_TAP(keycode) && record->tap.count) {
+        keycode = QK_LAYER_TAP_GET_TAP_KEYCODE(keycode);
+    } else if (IS_QK_MODS(keycode)) {
+        keycode = QK_MODS_GET_BASIC_KEYCODE(keycode);
+    } else if (IS_QK_ONE_SHOT_MOD(keycode)) {
+        keycode = 0xE0 + biton(QK_ONE_SHOT_MOD_GET_MODS(keycode) & 0xF) + biton(QK_ONE_SHOT_MOD_GET_MODS(keycode) & 0x10);
+    }
+    if (keycode > ARRAY_SIZE(code_to_name)) {
+        return;
+    }
+
+    // update keylog
+    key_name = pgm_read_byte(&code_to_name[keycode]);
+    last_row = record->event.key.row;
+    last_col = record->event.key.col;
+}
+
+static const char *depad_str(const char *depad_str, char depad_char) {
+    while (*depad_str == depad_char) {
+        ++depad_str;
+    }
+    return depad_str;
+}
+
+static void oled_render_keylog(void) {
+    const char *last_row_str = get_u8_str(last_row, ' ');
+    oled_write(depad_str(last_row_str, ' '), false);
+    oled_write("x", false);
+    const char *last_col_str = get_u8_str(last_col, ' ');
+    oled_write(depad_str(last_col_str, ' '), false);
+    oled_write(", k", false);
+    const char *last_keycode_str = get_u16_str(last_keycode, ' ');
+    oled_write(depad_str(last_keycode_str, ' '), false);
+    oled_write(":", false);
+    oled_write_char(key_name, false);
+}
+
+__attribute__((weak)) const char * get_layer_name_user(int layer) {
+    return "Unknown";
+}
+
+static void oled_render_layer(void) {
+    oled_write("Layer: ", false);
+    oled_write_ln(get_layer_name_user(get_highest_layer(layer_state)), false);
+}
+
+bool oled_task_kb(void) {
+    if (!oled_task_user()) {
+        return false;
+    }
+
+    oled_render_layer();
+    oled_render_keylog();
+    oled_write_ln("", false);
+    return false;
+}
+
+static void setupForFlashing(void) {
+    oled_clear();
+    oled_write("                     ", false);
+    oled_write("  In flash mode...   ", false);
+    oled_write("                     ", false);
+    oled_write("                     ", false);
+
+    // QMK is clever about only rendering a certain number of chunks per frame,
+    // but since the device will go into flash mode right after this call,
+    // we want to override this behavior and force all the chunks to be sent to
+    // the display immediately.
```

Comment:
just a heads up, on develop (which is to be merged at the end of the month... `oled_render_dirty(true)` will handle this without the extra code here.

Reply:
Added a todo with a commented out implementation. Wasn't sure if there was a better way?

Reply:
Yeah, the newer function was added explicitly for this sort of use case in mind.  Which is why I mentioned it.